sys/random/shaxprng.c

```diff
@@ -0,0 +1,200 @@
+#define STATE_SIZE                      (SHA1_DIGEST_LENGTH)
+typedef sha1_context shax_context_t;
+#elif IS_ACTIVE(MODULE_PRNG_SHA256PRNG)
+/* state size is digest length of SHA-256 */
+#define STATE_SIZE                      (SHA256_DIGEST_LENGTH)
+typedef sha256_context_t shax_context_t;
+#endif
+
+static inline void _shax_init(shax_context_t *ctx)
+{
+    if (IS_ACTIVE(MODULE_PRNG_SHA1PRNG)) {
+        sha1_init((sha1_context *)ctx);
+    }
+    else if (IS_ACTIVE(MODULE_PRNG_SHA256PRNG)) {
+        sha256_init((sha256_context_t *)ctx);
+    }
+}
+
+static inline void _shax_update(shax_context_t *ctx, const void *data, size_t len)
+{
+    if (IS_ACTIVE(MODULE_PRNG_SHA1PRNG)) {
+        sha1_update((sha1_context *)ctx, data, len);
+    }
+    else if (IS_ACTIVE(MODULE_PRNG_SHA256PRNG)) {
+        sha256_update((sha256_context_t *)ctx, data, len);
+    }
+}
+
+static inline void _shax_final(shax_context_t *ctx, void *digest)
+{
+    if (IS_ACTIVE(MODULE_PRNG_SHA1PRNG)) {
+        sha1_final((sha1_context *)ctx, digest);
+    }
+    else if (IS_ACTIVE(MODULE_PRNG_SHA256PRNG)) {
+        sha256_final((sha256_context_t *)ctx, digest);
+    }
+}
+
+/* allocate SHA context */
+static shax_context_t ctx;
+
+static uint32_t datapos = STATE_SIZE;
+static int8_t digestdata[STATE_SIZE];
+static int8_t prng_state[STATE_SIZE];
+
+void _updatestate(int8_t *state)
+{
+    signed last = 1;
+    char zf = 0;
+
+    for (int i = 0; i < STATE_SIZE; i++) {
+        signed v;
+        char t;
+        /* add two bytes */
+        v = (int)state[i] + (int)digestdata[i] + last;
+
+        /* result is lower 8 bits */
+        t = (char)(v & 0xff);
+
+        /* check for state collision */
+        zf = zf | (state[i] != t);
+
+        /* store result, */
+        state[i] = t;
+
+        /* high 8 bits are carry, store for next iteration */
+        last = (v >> 8);
+    }
+
+    /* make sure at least one bit changes! */
+    if (!zf) {
+       state[0]++;
+    }
+}
+
+void _random_bytes(uint8_t *bytes, size_t size) /* TODO: use with global API */
+{
+    uint32_t loc = 0;
+    while (loc < size)
+    {
+        int copy;
+        /* find min between remaining out bytes and so far unused seed bytes */
+        if ( (size-loc) < (STATE_SIZE - datapos) )
+        {
+            copy = size-loc;
+        }
+        else
+        {
+            /* in first iteration this will be 0 */
+            copy = STATE_SIZE - datapos;
+        }
+        if (copy > 0)
+        {
+            /* directly copy data to out buffer */
+            memcpy(bytes + loc, digestdata + datapos, copy);
+            datapos += copy;
+            loc += copy;
+        }
+        /* no out data ready, (re)fill internal buffer */
+        else
+        {
+            /* reset SHA internal state */
+            _shax_init(&ctx);
+
+            /* update SHA internal state with PRNG state */
+            _shax_update(&ctx, (void *)prng_state, sizeof(prng_state));
```

Comment:
Here and bellow: I think this cast is not needed
```suggestion
            _shax_update(&ctx, prng_state, sizeof(prng_state));
```